lib/llhttp/parser.js

```diff
@@ -0,0 +1,223 @@
+  let method;
+  let url;
+  let statusCode;
+  let statusMessage;
+  const upgrade = inst.exports.llhttp_get_upgrade(p);
+  const shouldKeepAlive = inst.exports.llhttp_should_keep_alive(p);
+
+  for (let c = 0; c < i[kHeadersFields].length; c++) {
+    rawHeaders.push(i[kHeadersFields][c], i[kHeadersValues][c])
+  }
+  i[kOnHeaders](rawHeaders);
+
+  if (type === HTTPParser.REQUEST) {
+    method = constants.METHODS[inst.exports.llhttp_get_method(p)];
+    url = i[kUrl];
+  } else if (type === HTTPParser.RESPONSE) {
+    statusCode = inst.exports.llhttp_get_status_code(p);
+    statusMessage = i[kStatusMessage];
+  }
+  return i[kOnHeadersComplete](versionMajor, versionMinor, rawHeaders, method,
+url, statusCode, statusMessage, upgrade, shouldKeepAlive);
+};
+
+const wasm_on_body = (p, at, length) => {
+  const i = instMap.get(p);
+  const body = Buffer.from(inst.exports.memory.buffer, at, length);
+  return i[kOnBody](body);
+};
+
+const wasm_on_message_complete = (p) => {
+  return instMap.get(p)[kOnMessageComplete]();
+};
+
+const instMap = new Map();
+
+const inst = new WebAssembly.Instance(mod, {
+  env: {
+    wasm_on_message_begin,
+    wasm_on_url,
+    wasm_on_status,
+    wasm_on_header_field,
+    wasm_on_header_value,
+    wasm_on_headers_complete,
+    wasm_on_body,
+    wasm_on_message_complete,
+  },
+});
+
+inst.exports._initialize(); // wasi reactor
+
+class HTTPParser {
+  static REQUEST = REQUEST;
+  static RESPONSE = RESPONSE;
+  static kOnMessageBegin = kOnMessageBegin;
+  static kOnHeaders = kOnHeaders;
+  static kOnHeadersComplete = kOnHeadersComplete;
+  static kOnBody = kOnBody;
+  static kOnMessageComplete = kOnMessageComplete;
+
+  constructor(type) {
+    this[kPtr] = inst.exports.llhttp_alloc(constants.TYPE[type]);
+    instMap.set(this[kPtr], this);
+
+    this[kUrl] = '';
+    this[kStatusMessage] = null;
+    this[kHeadersFields] = [];
+    this[kHeadersValues] = [];
+    this[kBody] = null;
+  }
+ 
+  [kReset]() {
+    this[kUrl] = '';
+    this[kStatusMessage] = null;
+    this[kHeadersFields] = [];
+    this[kHeadersValues] = [];
+    this[kBody] = null;
+  }
+
+  [kOnMessageBegin]() {
+    return 0;
+  }
+
+  [kOnHeaders](rawHeaders) {}
+
+  [kOnHeadersComplete](versionMajor, versionMinor, rawHeaders, method,
+    url, statusCode, statusMessage, upgrade, shouldKeepAlive) {
+    return 0;
+  }
+
+  [kOnBody](body) {
+    this[kBody] = body;
+    return 0;
+  }
+
+  [kOnMessageComplete]() {
+    return 0;
+  }
+
+  destroy() {
+    instMap.delete(this[kPtr]);
+    inst.exports.llhttp_free(this[kPtr]);
+  }
+
+  execute(data) {
+    // TODO(devsnek): could probably use static alloc and chunk but i'm lazy
+    const ptr = inst.exports.malloc(data.byteLength);
+    const u8 = new Uint8Array(inst.exports.memory.buffer);
+    u8.set(data, ptr);
+    const ret = inst.exports.llhttp_execute(this[kPtr], ptr, data.length);
+    inst.exports.free(ptr);
+    this[kCheckErr](ret);
+    return ret;
+  }
```

Comment:
FYI, this is where we are likely losing throughput.

Reply:
Yup, I agree.